### PR TITLE
alpine based building via xx

### PR DESCRIPTION
**Pull request by tonistiigi**
Update to alpine based building. Musl has much better static binaries support. This means ~riscv hacks and~ mips are removed. Could restore be mips but don't think anyone would be interested.

~32bit arm+386 not working~ atm. Not quite sure where the issue is. @crazy-max @AkihiroSuda 

```
In file included from config-temp/qemu-conf.c:1:
In file included from /usr/include/glib-2.0/glib.h:30:
In file included from /usr/include/glib-2.0/glib/galloca.h:32:
/usr/include/glib-2.0/glib/gtypes.h:463:3: error: '_GStaticAssertCompileTimeAssertion_0' declared as an array with a negative size
  G_STATIC_ASSERT(sizeof (unsigned long long) == sizeof (guint64));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gmacros.h:747:103: note: expanded from macro 'G_STATIC_ASSERT'
#define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                                                                      ^~~~~~~~~~~~~~~
config-temp/qemu-conf.c:7:4: error: 'qemu_build_bug_on' declared as an array with a negative size
   QEMU_BUILD_BUG_ON(sizeof(size_t) != GLIB_SIZEOF_SIZE_T);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
config-temp/qemu-conf.c:4:63: note: expanded from macro 'QEMU_BUILD_BUG_ON'
#define QEMU_BUILD_BUG_ON(x)   typedef char qemu_build_bug_on[(x)?-1:1] __attribute__((unused));
                                                              ^~~~~~~~
2 errors generated.



```

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

### CODE REVIEW

**Comment by tonistiigi** [comment]
Cross detection seems quite broken in qemu. When defining regular `--host/--cross-prefix` it doesn't compile with the correct architecture. The combination I found(at least for 64bit arches) was to set `--cc` to host architecture and `--build-cc` separately. I'm not even sure if this validation routine is supposed to run with host or build cc.

**Comment by tonistiigi** [comment]
32 bit systems building again now

File: docker-bake.hcl [hunk 1]
```diff
@@ -37,9 +37,7 @@ target "all-arch" {
     "linux/arm",
```
Comment (crazy-max) [inline docker-bake.hcl:37]: Default arm variant is v7. I think it should be distinguished (maybe include v6 as well).

```suggestion
    "linux/arm/v6",
    "linux/arm/v7",
```

Comment (tonistiigi) [reply docker-bake.hcl:37]: https://github.com/containerd/containerd/blob/master/platforms/platforms.go#L100 `linux/arm` is normalized. I don't really want to add v6 very much as there is no definition of what v6 means. Afaics, it means different things in alpine vs debian images. Better to let it die. Also, not related to this pr.

**Comment by tonistiigi** [comment]
I'll try to update this to keep riscv as there is some alpine support now.

**Comment by tonistiigi** [comment]
Building riscv binaries has been restored.

**Comment by tonistiigi** [comment]
hmm

```
#72 [linux/amd64 base 6/6] RUN set -e;   [ "$(xx-info arch)" = "ppc64le" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple;   [ "$(xx-info arch)" = "386" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple;   true
#72 sha256:0506d7ea10f531db73334f4685f045fd50967081c9a482d11568402af9689a62
#72 1.645 ln: /usr/bin/i586-alpine-linux-musl-ld: File exists
#72 ERROR: process "/bin/sh -c set -e;   [ \"$(xx-info arch)\" = \"ppc64le\" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple;   [ \"$(xx-info arch)\" = \"386\" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple;   true" did not complete successfully: exit code: 1
```

But I can't seem to repro locally.

**Comment by tonistiigi** [comment]
CI issue was caused by https://github.com/tonistiigi/xx/pull/24 . Didn't appear local because I'm on arm64.

File: .github/buildkit.toml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[worker.oci]
+  max-parallelism = 4
```
Comment (crazy-max) [inline .github/buildkit.toml:2]: Any improvements if we fix parallelism?

File: Dockerfile [hunk 3]
```diff
@@ -1,85 +1,52 @@
...
+  [ "$(xx-info arch)" = "ppc64le" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple; \
+  [ "$(xx-info arch)" = "386" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple; \
+  true
 
-FROM base-$TARGETARCH$TARGETVARIANT AS base
 
 FROM base AS build
 ARG TARGETPLATFORM
-ARG QEMU_VERSION
+ARG QEMU_VERSION QEMU_TARGETS
```
Comment (crazy-max) [inline Dockerfile:43]: Should not be?:

```suggestion
ARG QEMU_VERSION
ARG QEMU_TARGETS
```

File: .github/buildkit.toml (hunk 2 above)
Comment (tonistiigi) [reply .github/buildkit.toml:2]: The parallelism goes so big it sometimes runs out of memory without it if I remember correctly. Also faster

File: Dockerfile (hunk 3 above)
Comment (tonistiigi) [reply Dockerfile:43]: What's the difference?

Comment (crazy-max) [reply Dockerfile:43]: I didn't think this syntax could work but apparently it does!

**Review by crazy-max — APPROVED** [review]

**Pull request merged**